### PR TITLE
Add Search support

### PR DESCRIPTION
**Pull request by willholley**
## Overview
Adds a search module to Fauxton. The functionality is only enabled
upon detection of "search" in the reported CouchDB features.

When enabled, it adds:
 * New dropdown options to create/update search indexes in the sidebar
 * New panel to run search queries from the sidebar
 * Text index templates to the Mango Index editor

## Testing recommendations

Using CouchDB with clouseau (e.g. using the Helm chart), verify that you can:
 * Create a new search index
 * Update a search index
 * Query a search index
 * Create a new Mango text index

From the UI.

For CouchDB 2 / master without clouseau deployed, there should be no options for search indexes in the UI and no errors.

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [x] Documentation reflects the changes;
- [x] Update [rebar.config.script](https://github.com/apache/couchdb/blob/master/rebar.config.script) with the correct tag once a new Fauxton release is made

### CODE REVIEW

**Comment by willholley** [comment]
cc @Antonio-Maranhao

File: app/addons/search/components/Analyzer.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    singleAnalyzer: PropTypes.string.isRequired,
+    setAnalyzerType: PropTypes.func.isRequired,
+    setSingleAnalyzer: PropTypes.func.isRequired,
+    setDefaultMultipleAnalyzer: PropTypes.func.isRequired,
+    addAnalyzerRow: PropTypes.func.isRequired
+  };
+
+  constructor(props) {
+    super(props);
+    this.selectAnalyzerType = this.selectAnalyzerType.bind(this);
```
Comment (popojargo) [inline app/addons/search/components/Analyzer.js]: Why binding to this when using arrow functions?

**Comment by willholley** [comment]
Thanks for the great review @popojargo. I've updated the PR with your suggestions.

**Comment by popojargo** [comment]
@willholley Seems to be few errors with Travis. Also I won't approve since I can't really test the Clouseau Setup locall.y I assume Antonio will be able to test it out :)

**Comment by willholley** [comment]
@popojargo if you or @Antonio-Maranhao want to test, there is a Docker container at `ibmcom/couchdb3:preview` which runs CouchDB/master and `clouseau`.

**Comment by willholley** [comment]
The nightwatch tests are a bit tricky given they will only work when search is enabled. I can temporarily point them to the Dockerfile mentioned above but we may want to consider a dedicated CI container and/or include clouseau as an option in the semi-official CouchDB 3 container.

**Comment by willholley** [comment]
my inclination is just to remove the failing nightwatch tests for now

**Comment by popojargo** [comment]
What if travis runs a matrix of 2 jobs.

- One with the latest of CouchDB: Runs stylecheck, jest and nightwatch but exclude Search addon test suite
- One with a Clouseau image that only run the specific nightwatch test suite?

**Comment by willholley** [comment]
@popojargo we could do that y - have a couchdb2 and couchdb3 job (which uses the `ibmcom` image for now). I'll investigate...

**Comment by willholley** [comment]
nightwatch tests now passing with a combined CouchDB / master + clouseau container.

**Comment by popojargo** [comment]
cc: @wohali  

For the moment I assume it's fine to run all the Fauxton tests on the IBM Couch image but we should eventually use either an official Apache-couchdb image or only use the IBM image for specific IBM features.

I'll try to test this pr today so you can merge it tomorrow!

**Comment by willholley** [comment]
@popojargo agree it's not ideal to use the IBM image but afaik we don't have a semi-official container that represents CouchDB master. There's nothing IBM-specific in the `ibmcom/couchdb3:preview` container so this could easily be switched to use a CouchDB 3 / master container when available. We could also choose to build / host a container specifically for Fauxton CI.

I'm looking at adding a CouchDB 2/3 build matrix but need to figure out how to conveniently disable tests that don't apply to CouchDB 2 and/or CouchDB 3 with Search.

**Comment by wohali** [comment]
@popojargo @willholley because of licensing issues I don't think we can release an Apache couchdb image with search in it. You'll have to do something like `docker-compose` to include search anyway.

Yes, we're not allowed to push images to Docker (either under `apache/couchdb` or `couchdb` top level) that represent an unreleased version of CouchDB, sorry.

**Comment by willholley** [comment]
this is a bit of a rabbit hole due to the old version of nightwatch that the test suite currently uses not supporting tags. I'm investigating whether we can trigger particular tests using some other mechanism (e.g. conditionally run based on advertised CouchDB features).

**Comment by popojargo** [comment]
You could make an helper function that skipTestUnless('<feature>') is enabled. That function could be called in the **before** hook. You could also cache the features list to avoid querying the API for features list everytime.

**Comment by wohali** [comment]
One idea, if you would rather use the Apache docker image (I'd prefer it frankly) would be to clone the `apache/couchdb-docker` repo, then build the `dev` Docker image from it, and then use that to run your tests.

Would add about 5-10 mins to your build time, which shouldn't be horrible.

**Comment by willholley** [comment]
The `dev` Dockerfile currently fails to run the tests. The CouchDB error is:

```
[error] 2019-10-03T13:41:13.642716Z nonode@nohost <0.2750.0> -------- gen_server ioq_server terminated with reason: bad argument in call to erlang:monitor(process, {main,'clouseau@127.0.0.1'}) at ioq_server:submit_request/2(line:417) <= ioq_server:maybe_submit_request/1(line:342) <= ioq_server:handle_info/2(line:171) <= gen_server:try_dispatch/4(line:601) <= gen_server:handle_msg/5(line:667) <= proc_lib:init_p_do_apply/3(line:247)
  last msg: timeout
     state: {state,188478,ioq_histos,[],20,{[{channel,other,{[{request,{main,'clouseau@127.0.0.1'},version,other,other,{<0.2986.0>,#Ref<0.0.2.4717>},undefined,-576460635523539056,undefined}],[]},{[],[]},{[],[]}}],[]},{[],[]},{[],[]},{[],[]},true,[1.0,0.001,0.0001,0.0001],[1.0,1.0,1.0]}
    extra: []
[error] 2019-10-03T13:41:13.643067Z nonode@nohost <0.2750.0> -------- CRASH REPORT Process ioq_server (<0.2750.0>) with 0 neighbors exited with reason: bad argument in call to erlang:monitor(process, {main,'clouseau@127.0.0.1'}) at ioq_server:submit_request/2(line:417) <= ioq_server:maybe_submit_request/1(line:342) <= ioq_server:handle_info/2(line:171) <= gen_server:try_dispatch/4(line:601) <= gen_server:handle_msg/5(line:667) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {ioq_server,init,['Argument__1']}, ancestors: [ioq_sup,<0.216.0>], messages: [], links: [<0.218.0>], dictionary: [{rand_seed,{#{max => 288230376151711743,next => #Fun<rand.8.41921595>,...},...}}], trap_exit: false, status: running, heap_size: 1598, stack_size: 27, reductions: 67516
[error] 2019-10-03T13:41:13.643394Z nonode@nohost <0.218.0> -------- Supervisor ioq_sup had child ioq_server started with ioq_server:start_link() at <0.2750.0> exit with reason bad argument in call to erlang:monitor(process, {main,'clouseau@127.0.0.1'}) at ioq_server:submit_request/2(line:417) <= ioq_server:maybe_submit_request/1(line:342) <= ioq_server:handle_info/2(line:171) <= gen_server:try_dispatch/4(line:601) <= gen_server:handle_msg/5(line:667) <= proc_lib:init_p_do_apply/3(line:247) in context child_terminated
[error] 2019-10-03T13:41:13.643507Z nonode@nohost <0.218.0> -------- Supervisor ioq_sup had child ioq_server started with ioq_server:start_link() at <0.2750.0> exit with reason reached_max_restart_intensity in context shutdown
[notice] 2019-10-03T13:41:13.643975Z nonode@nohost <0.2986.0> 45e54b7e8f localhost:5984 172.24.0.1 undefined GET / 200 ok 3
[info] 2019-10-03T13:41:13.653604Z nonode@nohost <0.9.0> -------- Application ioq exited with reason: shutdown
[error] 2019-10-03T13:41:13.668074Z nonode@nohost <0.543.0> -------- gen_server <0.543.0> terminated with reason: killed
  last msg: {'EXIT',<0.407.0>,killed}
     state: {state,163961,couch_replicator_doc_processor,nil,<<"_replicator">>,#Ref<0.0.2.1253>,nil,[],true}
    extra: []
[error] 2019-10-03T13:41:13.668405Z nonode@nohost <0.543.0> -------- CRASH REPORT Process  (<0.543.0>) with 0 neighbors exited with reason: killed at gen_server:terminate/7(line:812) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {couch_multidb_changes,init,['Argument__1']}, ancestors: [<0.407.0>,couch_replicator_sup,<0.381.0>], messages: [], links: [], dictionary: [], trap_exit: true, status: running, heap_size: 6772, stack_size: 27, reductions: 26261
[os_mon] memory supervisor port (memsup): Erlang has closed
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
Kernel pid terminated (application_controller) ({application_terminated,ioq,shutdown})

{"Kernel pid terminated",application_controller,"{application_terminated,ioq,shutdown}"}
```

**Comment by wohali** [comment]
@willholley Can you get someone more familiar with the Clouseau code to help with this? Granted, our `dev` Docker container won't include the additional clouseau binaries, if these tests are intended to require those, you need a `docker-compose` approach that hopefully someone out there can explain.

**Comment by willholley** [comment]
https://github.com/apache/couchdb/pull/2229 should fix the above error. The CouchDB dev container doesn't currently have an option to include clouseau so I've kept the IBM container in the matrix for now.

**Comment by willholley** [comment]
confirmed that re-running the tests in Travis against CouchDB master is working, so this should be good to merge.

**Comment by willholley** [comment]
good spot @popojargo - pushed a fix.

**Comment by Antonio-Maranhao** [comment]
Great work @willholley

**Pull request merged**